### PR TITLE
🔭 [GRPO] Log advantages and fraction of samples with an std of zero

### PR DESCRIPTION
**Pull request by edbeeching**
# What does this PR do?
This PR implements:

- Logging of advantages in the wandb table
- Logging the fraction of samples which have a reward std of zero.

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -1239,10 +1241,11 @@ def _generate_and_score_completions(
             self._metrics[mode][f"rewards/{reward_func_name}/std"].append(std_rewards)
         self._metrics[mode]["reward"].append(mean_grouped_rewards.mean().item())
         self._metrics[mode]["reward_std"].append(std_grouped_rewards.mean().item())
-
+        self._metrics[mode]["frac_zero_reward_std"].append(frac_zero_std.float().mean().item())
```
Comment (lewtun) [inline trl/trainer/grpo_trainer.py:1244]: To avoid confusion with users thinking this refers to "zero reward" instead of "rewards with zero std dev", let's rename to this:
```suggestion
        self._metrics[mode]["frac_reward_zero_std"].append(frac_zero_std.float().mean().item())
```

**Review by lewtun — APPROVED** [review]
LGTM with a nit and once the docs are updated too: https://github.com/huggingface/trl/blob/main/docs/source/grpo_trainer.md#logged-metrics

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -1239,10 +1241,11 @@ def _generate_and_score_completions(
...
         self._metrics[mode]["reward"].append(mean_grouped_rewards.mean().item())
         self._metrics[mode]["reward_std"].append(std_grouped_rewards.mean().item())
-
+        self._metrics[mode]["frac_reward_zero_std"].append(frac_zero_std.float().mean().item())
         # Log prompt and completion texts
         self._textual_logs["prompt"].extend(gather_object(prompts_text))
         self._textual_logs["completion"].extend(gather_object(completions_text))
+        self._textual_logs["advantages"].extend(all_process_advantages.tolist())
         for i, name in enumerate(self.reward_func_names):
             self._textual_logs["rewards"][name].extend(rewards_per_func[:, i].tolist())
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1251]: just re-ordering:

```suggestion
        for i, name in enumerate(self.reward_func_names):
            self._textual_logs["rewards"][name].extend(rewards_per_func[:, i].tolist())
        self._textual_logs["advantages"].extend(all_process_advantages.tolist())
```

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -1239,10 +1241,11 @@ def _generate_and_score_completions(
             self._metrics[mode][f"rewards/{reward_func_name}/std"].append(std_rewards)
         self._metrics[mode]["reward"].append(mean_grouped_rewards.mean().item())
         self._metrics[mode]["reward_std"].append(std_grouped_rewards.mean().item())
-
+        self._metrics[mode]["frac_reward_zero_std"].append(frac_zero_std.float().mean().item())
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1244]: nit

```suggestion
        self._metrics[mode]["frac_reward_zero_std"].append(frac_zero_std.float().mean().item())

```

**Comment by qgallouedec** [comment]
Now it looks good!

Screenshot 2025-05-27 at 12 46 37

**Comment by qgallouedec** [comment]
And the new metric:

Screenshot 2025-05-27 at 12 55 39

```python
import tempfile
from datasets import load_dataset
from trl import GRPOConfig, GRPOTrainer
import random

config_name = "standard_prompt_only"
dataset = load_dataset("trl-internal-testing/zen", config_name, split="train")

def random_reward(completions, **kwargs):
    return [random.choice([0.0, 1.0]) for _ in completions]

with tempfile.TemporaryDirectory() as tmp_dir:
    training_args = GRPOConfig(
        output_dir=tmp_dir,
        learning_rate=0.1,  # increase the learning rate to speed up the test
        per_device_train_batch_size=6,  # reduce the batch size to reduce memory usage
        num_generations=3,  # reduce the number of generations to reduce memory usage
        max_completion_length=8,  # reduce the completion length to reduce memory usage
        log_completions=True,
        logging_steps=3,
    )
    trainer = GRPOTrainer(
        model="trl-internal-testing/tiny-Qwen2ForCausalLM-2.5",
        reward_funcs=random_reward,
        args=training_args,
        train_dataset=dataset,
    )

    trainer.train()
```
